lib/ecto/adapters/sql.ex

```diff
@@ -934,7 +935,8 @@ defmodule Ecto.Adapters.SQL do
   @doc false
   def reduce(adapter_meta, statement, params, opts, acc, fun) do
     %{pid: pool, telemetry: telemetry, sql: sql, opts: default_opts} = adapter_meta
-    opts = with_log(telemetry, params, opts ++ default_opts)
+    log_params = opts[:cast_params] || params
+    opts = with_log(telemetry, log_params, opts ++ default_opts)
```

Comment:
I think we need to be careful here because this will change the `params` we send in the telemetry event, no? It is probably best to move the `opts[:cast_params] || params` bit to inside `with_log`. :)